funding/manager.go

```diff
@@ -1801,17 +1809,17 @@ func (f *Manager) fundeeProcessOpenChannel(peer lnpeer.Peer,
 	}
 
 	if resCtx.reservation.IsTaproot() {
-		if msg.LocalNonce == nil {
-			err := fmt.Errorf("local nonce not set for taproot " +
-				"chan")
-			log.Error(err)
-			f.failFundingFlow(
-				resCtx.peer, cid, err,
-			)
+		localNonce, err := msg.LocalNonce.UnwrapOrErrV(errNoLocalNonce)
+		if err != nil {
```

Comment:
i find the `UnwrapOrErr` a bit strange since it only ever can return the error that it is given. So wont it make more sense for it to return a boolean to indicate the same thing?

Reply:
There's nothing wrong with returning the argument unchanged. It reduces the amount of usage context required. If you want to insist on it being a boolean (which btw I think [booleans should generally be avoided if at all possible[1]](https://existentialtype.wordpress.com/2011/03/15/boolean-blindness/), then you would want a function like [note](https://hackage.haskell.org/package/errors-2.3.0/docs/Control-Error-Util.html#v:note) that can inject an Option into Either space. Since go is... go, we don't really use Either's for error reporting and so this would be the next closest thing.

> [1] Another harm is the condition of Boolean blindness alluded to earlier.  Suppose that I evaluate the expression e=e’ to test whether e and e’ are equal.  I have in my hand a bit.  The bit itself has no intrinsic meaning; I must associate a provenance with that bit in order to give it meaning.  “This bit being true means that e and e’ are equal, whereas this other bit being false means that some other two expressions are not equal.”  Keeping track of this information (or attempting to recover it using any number of program analysis techniques) is notoriously difficult.  The only thing you can do with a bit is to branch on it, and pretty soon you’re lost in a thicket of if-the-else’s, and you lose track of what’s what.  Evolve the program a little, and you’re soon out to sea, and find yourself in need of sat solvers to figure out what the hell is going on.

I really want to encourage us to think more composably when it comes to writing code. Composability is what gives us leverage. What makes things composable is making components that can be reused like this and not requiring everyone do their own type marshalling. I think this function gives us a valuable tool. It's less valuable in go than in other language ecosystems but that doesn't mean it doesn't give us any value. IIRC @ellemouton you said you have an extension that hides the `if err != nil { return _, err }` from view. Choosing a boolean here may break that, no?

Reply:
> i find the UnwrapOrErr a bit strange since it only ever can return the error that it is given. So wont it make more sense for it to return a boolean to indicate the same thing?

Why do you think the bool is better? After working with the new package for a bit, I think most of the time, this new method is the best way to: force a nil-ness check, then bail out if things are `None`. It also ends up (IMO) being the most Go-like way to unwrap something and handle the error case. The pattern also is conducive to testing as unit tests can assert a given error is returned once a certain field is blank. 

If we just returned the bool, we'd _still_ want to return an error, no/

Reply:
> It reduces the amount of usage context required. If you want to insist on it being a boolean

I dont really agree. To me, if a func returns an error, it means something went wrong in that function and we shouldnt make assumptions about what went wrong without checking exactly what error was returned. 

Whereas the `thing, ok := maybeGetThing()` is a common pattern in go & it is clear that the `ok` boolean is saying something about if `thing` is populated or not. 

> IIRC @ellemouton you said you have an extension that hides the if err != nil { return _, err } from view.

nope. I dont hide it. 

> Why do you think the bool is better?

see above reply re an error being returned from a func meaning that something went wrong in that func. 

> The pattern also is conducive to testing as unit tests can assert a given error is returned once a certain field is blank.

Yes I like the `UnwrapOrErr(t *testing)` helper but I feel like that is orthogonal to this. 

> If we just returned the bool, we'd still want to return an error, no?

Yes, but I think it is important that that error be derived on the correct layer. For me, the 
```
localNonce, err := msg.LocalNonce.UnwrapOrErrV(errNoLocalNonce)
if err != nil {
```

makes it look as though things errored out on a layer below this method. Ie, it isnt clear that we are actually checking in the context of this function that: "if taproot channel & local nonce not set then _this_ function should error now". I think the boolean way makes this more clear.

Reply:
> said you have an extension that hides the if err != nil { return _, err } from view. Choosing a boolean here may break that, no?

Btw - I think in general it only makes sense to hide this because it means that something went wrong downstream of this func. 

whereas here, things go wrong _here_. I think we dont want to hide why we are deciding to error out here if the decision is being made here. We want to be able to read "if taproot & no nonce, return an error"

Reply:
I tend to agree with @ellemouton here as (to me at least) what would be more natural and idiomatic Go is if `Option` would have a bare `Unrwap()` function that would return `(T, bool)`. To me "flowing" the error through the unwrap feels unnecessary and doesn't add much value as we're already checking if the returned error is `nil`. Instead we can just check `if !ok` (which is equally or more readable) and return the context specific error from the scope defined by the `if`. Just my two sats here, as I see value in API richness and composability too but I'd always choose simplicity over "complexity" (however small it is) as it feels like the unwrap patterns is going to spread all over our code base.

Reply:
One more comment on this is that if the error is inside the `if`'s scope then we can always add more context without compromising readability (so wrap with `fmt.Errorf(...)` or similar).

Reply:
> I dont really agree. To me, if a func returns an error, it means something went wrong in that function and we shouldnt make assumptions about what went wrong without checking exactly what error was returned.

Not sure what you mean in that you can always just wrap the error. This is the case of being passed a `nil` pointer when you want an actual value. I think you should view this just as a wrapper around a `ptr != nil` check. 

As mentioned above, this is a slightly less safe version than the existing methods as if the caller fails to check the error, you _can_ end up with a `nil`/empty value, so the map and `WhenSome` variants are to be preferred in most cases. This was made for convenience in tests and to make control flow a bit more linear. 

The main advantage of using the option wrapper at all is that you're _forced_ to do this check, and reviewers of a PR don't need to meticulously scan for potential nil pointer derefs to the same degree as before.